### DISCUSSION TITLE
Several Jamulus musicians in one place, benefit multichannel audio interface for Jamulus ... so far together only under one Jamulus participant then the mix online available... we need for each audio interface participant an own Jamulus participant

### DISCUSSION THREAD

**Discussion by dh1sa** [jamulussoftware/jamulus#2268] [Technical Support & Help]
ENGLISH:
Hello,

we use Jamulus to make music together with our band online. 
4 band members are in one room and play together there. For this we use a multichannel audio interface. 
With Jamulus, all 4 band members always come under one Jamulus participant. This is bad, because the others then can not mix them.
Is there no possibility to connect several channels of an audio interface to several Jamulus participants.
On the MAC or PC, the audio interface logs in with 16 input channels.
Perfect would be if the 4 of them are then each 4 Jamulus participants, so that these can then be mixed by the others.

4 PCs and 4 audio interface would work, 4 Internet accesses, but is not a solution. Too complex, too complicated. 
And about the 16 channel audio interface runs namely also the local monitoring. Jamulus gets an input here so that all 4 also hear the Jamulus Internet signal. Since we have little latency, the 4 hear each other directly, so not from the Jamulus server.

GERMAN:
Hallo,

wir nutzen Jamulus um mit unserer Band online zusammen Musik zu machen. 
4 Bandmitglieder sind in einem Raum und spielen dort zusammen. Dazu nutzen wir ein Mehrkanal Audio Interface. 
Bei Jamulus kommen damit immer alle 4 Bandmitglieder unter einem Jamulus Teilnehmer. Das ist schlecht, da die anderen online diese dann nicht mehr Mischen können.
Gibt es denn keine Möglichkeit mehrere Kanäle eines Audiointerfaces auf mehrere Jamulus Teilnehmer zu verbinden.
Am MAC oder PC meldet sich das Audio - Interface mit 16 Eingangskanäle an.
Perfekt wäre wenn die 4 davon dann jeweils 4 Jamulus Teilnehmer sind, so dass diese dann von den anderen gemischt werden können.

4 PCs und 4 Audiointerface würden funktionieren, 4 Internet Zugänge, ist aber keine Lösung. Zu aufwendig, zu kompliziert. 
Und über das 16 Kanal  Audio Interface läuft nämlich auch das lokale Monitoring. Jamulus bekommt hier einen Eingang so dass alle 4 auch das Jamulus Internetsignal hören. Da wir wenig Latenz haben, hören die drei sich direkt, also nicht vom Jamulus Server.

**Comment by gilgongo** [comment]
Jamulus lets you run multiple clients on the same machine. So, using a single PC, would you be able to have separate clients connected to separate outputs from your audio interface?  Each output would be connected to an input from a different musician in the room. 

> the 4 hear each other directly, so not from the Jamulus server.

That might cause you problems keeping in time with other players on the server. Better to have everyone listening to the sound that comes back from server (so, using headphones).

**Reply by dh1sa** [reply]
Hello, thanks for the answers!

I downloaded the linked document and used it. 
I use extra as in the instructions a MAC. MAC OS is Monterey.

Have duplicated in the Finder Jamulus. Gave each duplicate a different name. I can open these and change the setting on each. This works great!!! I can use several parallel Jamulus clients on different sound cards - USB - ports.

Now to my problem that I can't get solved. I am in despair, already put a lot of time into it.
If I end the individual Jamulus applications, as I understand it in all Jamulus applications the last settings of the Jamulus application is taken over, where the settings were changed.
Again better explained: All applications have then at the next start exactly the same settings - parameters. I must then always adjust by hand. Costs a lot of time and nerves with four clients.
You write there from the call with "--inifile". Exactly this Inifile I do not find also do not get the call.

Still to the other remarks:
We use our own Jamulus server. It is a Linux server, located at the same site and runs on separate hardware.
We all play with headphones, use as monitor "Ultranet" of the mixer. Ultranet gives all input channels of the mixer digitally to the monitors, which are exclusively headphones for us. The whole thing works great. The mixer is a MIDAS MR 18. I can only recommend for Jamulus, he has 16 USB sound card - channels. So 16 digital inputs and 16 digital outputs.
We are a Christian pop band and use Jamulus for band rehearsals for a long time. Band members live far geographically scattered. Since electric bass, electric guitar and keyboard as well as vocals come from the rehearsal room, we need 4 Jamulus single chairs so that others can mix. You know that, the drummer wants much louder the electric bass... etc.
Many greetings Axel and maybe someone can help me with the problem.

**Comment by chrisrimple** [comment]
I have [some instructions here](https://docs.google.com/document/d/1smcvsxdaaViPQvGMQHmah_6BQeqowhmGSFMHfnlY2FI/) for how to run multiple Jamulus clients.  But you will definitely need each person to wear headphones.

**Reply by dh1sa** [reply]
Hello, thanks for the answers!

I downloaded the linked document and used it. 
I use extra as in the instructions a MAC. MAC OS is Monterey.

Have duplicated in the Finder Jamulus. Gave each duplicate a different name. I can open these and change the setting on each. This works great!!! I can use several parallel Jamulus clients on different sound cards - USB - ports.

Now to my problem that I can't get solved. I am in despair, already put a lot of time into it.
If I end the individual Jamulus applications, as I understand it in all Jamulus applications the last settings of the Jamulus application is taken over, where the settings were changed.
Again better explained: All applications have then at the next start exactly the same settings - parameters. I must then always adjust by hand. Costs a lot of time and nerves with four clients.
You write there from the call with "--inifile". Exactly this Inifile I do not find also do not get the call.

Still to the other remarks:
We use our own Jamulus server. It is a Linux server, located at the same site and runs on separate hardware.
We all play with headphones, use as monitor "Ultranet" of the mixer. Ultranet gives all input channels of the mixer digitally to the monitors, which are exclusively headphones for us. The whole thing works great. The mixer is a MIDAS MR 18. I can only recommend for Jamulus, he has 16 USB sound card - channels. So 16 digital inputs and 16 digital outputs.
We are a Christian pop band and use Jamulus for band rehearsals for a long time. Band members live far geographically scattered. Since electric bass, electric guitar and keyboard as well as vocals come from the rehearsal room, we need 4 Jamulus single chairs so that others can mix. You know that, the drummer wants much louder the electric bass... etc.
Many greetings Axel and maybe someone can help me with the problem.

**Reply by gilgongo** [reply]
Yes, you need to start each client with it's own inifile. So as per the [documentation here](https://jamulus.io/wiki/Software-Manual#command-line-options). To find your current ini file location, run `find ~ -name Jamulus.ini -ls` 

Also BTW if you are using the current signed Jamulus macOS release, your ini files can only be in the directory where your current one is.

**Reply by pgScorpio** [reply]
But still this is not possible with two participants on a two channel interface, since with a two channel interface Jamulus doesn't allow to select different input channels in the two instances.

Is there already a change request for this "bug"???
This problem is also mentioned in several other threads, but I can't see a change/pull request for this issue, so I don't know if it is already being solved or not.

**Reply by ghost** [reply]
A 2 channel interface is a Stereo interface (right down to the hardware) for inputs and outputs and microphones and headphones. This does not prevent a stereo interface from being used as a 2 channel interface, but may explain why the interface is assumed to be stereo. Yes, your laptop that has a microphone beside the screen's camera, and has a mono microphone on the TRRS headhone jack is using a Stereo interface from the hardware codec's viewpoint. The integrated circuit is stereo in (microphones) and stereo out (headphones).

**Reply by dcorson-ticino-com** [reply]
> Is there already a change request for this "bug"???

This is not a bug, it is part of the specification of the program.
If you need separate channels, start multiple instances of the program.
There has often been discussion to allow the two mono input channels to act as two different users and no satisfying solution has been found.  I have spent a lot of time trying to get one of the suggested solution paths to work, but gave up for the moment as the program just wasn't planned to do this.   A PR exists if you look, but it is not satisfactory with several problems as yet without solution.

If you have a better solution, then I would suggest that you start coding.  You do realize that this is a free program that people make in their spare time for fun.  Each person contributing the solution to the problem that is their particular bugaboo.  If this is yours, go to it!

**Reply by pgScorpio** [reply]
> This is not a bug, it is part of the specification of the program.

That's why "bug" is quoted, it's a strange limitation by decision, not by the code libraries. When I have a four channel interface I can run four clients, each using the same single channel for both inputs, but on a stereo interface I can't run two clients each using a single channel. Same goes for not accepting a mono interface. No real limitation for that too, since it would still be possible to select the same input for left and right.

> If you have a better solution, then I would suggest that you start coding

I would love to, if I had the time. And there are a lot of people already more familiar with the code, so they probably come up with a better/faster solution (as far as I can see it is mainly some definitions and deleting/adjusting some checks).

The main problem here is not the coding, but the decission making...

**Reply by pgScorpio** [reply]
> A 2 channel interface is a Stereo interface (right down to the hardware)

Is that so?
Just take a Behringer UMC 22 as an example.
One mic input and one instrument input is two channel, but not neccesarily stereo!

> This does not prevent a stereo interface from being used as a 2 channel interface,

So, why does Jamulus actually do prevent using it as a two channel interface?

> The integrated circuit is stereo in (microphones) and stereo out (headphones).

Also not true in many cases...
Most of the time the mic hardware is mono, but handled as pseudo stereo by the driver.

**Reply by dcorson-ticino-com** [reply]
> as far as I can see it is mainly some definitions and deleting/adjusting some checks).

I guess that means you haven't looked t the code at all.....

**Reply by pgScorpio** [reply]
Wrong guess 😉
But indeed, as said, I don't have the time, so I might have missed something. So others, more familiar with the code, should have a better look perhaps?

**Reply by ghost** [reply]
> > A 2 channel interface is a Stereo interface (right down to the hardware)
> 
> Is that so? Just take a Behringer UMC 22 as an example. One mic input and one instrument input is two channel, but not neccesarily stereo!

I'm Not talking about some add-on, I'm talking about what Is In your computer:
https://www.ti.com/lit/gpn/tlv320aic3204
> 
> > This does not prevent a stereo interface from being used as a 2 channel interface,
> 
> So, why does Jamulus actually do prevent using it as a two channel interface?
> 
Jamulus didn't get around to adding extra time critical code for something you can purchase as an add-on.
> > The integrated circuit is stereo in (microphones) and stereo out (headphones).
> 
> Also not true in many cases... Most of the time the mic hardware is mono, but handled as pseudo stereo by the driver.

That is what the external shape of your laptop (and GUI) may cause you to think. Disassemble your laptop to find the truth, and also see the Stereo Codec data sheet. e.g.
https://www.ti.com/lit/gpn/tlv320aic3204

**Reply by pgScorpio** [reply]
@DavidSavinkoff 
I guess we are on completely different wavelengths, So I don't know if we should discuss this any further here, but...

I'm not talking about PC audio, but about the USB interfaces which are mostly used with Jamulus.
And the very often used cheaper ASIO ones are not stereo, but have 2 different channels: mic and instrument, so it would also be very logical to use two clients with these.

Also there are many different audio codecs, and the ti version you mention is certainly not in my laptop, (as most PC's and Laptops use Realtek or Intel codecs, and yes the modern ones are all stereo.) But you better check what chipsets are used in the cheap USB headsets that are also very often used with Jamulus too (many of them DO use mono codecs, and still are 2 channel in windows. If not you can't use them for Jamulus, and also no real good reason why this wouldn't work!). 

And about "adding time critical code". Did you add time critical code for interfaces with more than 2 channels??? Since with these I CAN select 1 same input for both Jamulus channels! (afaik, no special code involved, just two times the same channel index, so even for mono devices you could always use 2x index 0)

Please think in solutions, and not in problems...

**Reply by ghost** [reply]
The datasheet is Texas Instruments and the codec is a cloned Realtek (Realtek ALC3204) used everywhere.
Jamulus has stereo which is better than mono. A mono Jamulus has no benefits over stereo because of the way Jamulus is designed. See the mono and stereo options: Settings > Audio/Network Setings > Audio Channels
Jamulus also lets you run multiple clients on the same machine without jamming the network. Problem Solved.

**Reply by ghost** [reply]
@pgScorpio
> But still this is not possible with two participants on a two channel interface, since with a two channel interface Jamulus doesn't allow to select different input channels in the two instances.

There is a solution fo this if you use Jack (QjackCtl in Linux :: Linux Rules!):
Connect two stereo Jamulus client instances to a Jamulus server.
Execute the QjackCtl program and click on the Connect button.
Disconnect the rats nest of connections.
Connect:
system capture_1 to jamulus input left
system capture_2 to jamulus-01 input left
jamulus output left to system playback_1
jamulus-01 output left to system playback_2
(I think 'left' is mono. Use 'right' if left doesn't work)

Note that the simple nature of Jamulus allows you to do amazing things without adding bloated features.

**Reply by pgScorpio** [reply]
So indeed we are on diifferent wavelenghts...

> The datasheet is Texas Instruments and the codec is a cloned Realtek (Realtek ALC3204) used everywhere.

As already said, I was NOT addressing computer audio codecs....

> Jamulus has stereo which is better than mono. A mono Jamulus has no benefits over stereo because of the way Jamulus is designed

Just because Jamulus was designed this way does not mean stereo is better than mono. Stereo is actually always created from two or more mono channels!
Note that every (Semi-)professional mixer is primarily mono-in stereo-out. With good reason!
And, though Jamulus has a mono-in/stereo-out mode, in Jamulus a real (actual pseudo) mono input is only possible with interfaces that have more than 2 inputs?? I really call this a "bug" (Note the quotes again, since it's actually just a bad decision.).

> There is a solution fo this...

Yes I very well know there are several solutions by adding software/hardware, but the point is this is far to advanced/expensive for the average user and should not be necessary if Jamulus had it's basics right.


> Note that the simple nature of Jamulus allows you to do amazing things without adding bloated features.

Yes Jamulus is amazing, but this is not a bloated feature. It is just basics. And in my opinion one should solve all basic issues first, before implementing advanced ("bloated??") features (like "Auto set Faders".).

 
And again: Should we really discuss this in this topic????

**Reply by hoffie** [reply]
> but this is not a bloated feature. It is just basics. And in my opinion one should solve all basic issues first, before implementing advanced ("bloated??") features (like "Auto set Faders".).

I understand that support for this scenario is beneficial in certain environments and it sounds like a simple feature or extension. In fact, it might have been simple if it was built that way in the first place. However, it wasn't. My gut feeling is that this is not a trivial extension at all, because it either requires a single Jamulus instance to behave like multiple instances on a protocol level (i.e. handling multiple streams) or because the protocol and/or server would require support for more than two channels in order to signal that back to clients. In any case, it would need UI changes.
It certainly should be possible to build this. I think @dcorson-ticino-com tried with some success and is therefore probably rather qualified to judge the amount of work for this to be implemented.

As to why other features which may not sound as basic as this one to you are implemented while this isn't: I guess it boils down
- to code/architecture simplicity (e.g. *Auto set faders* is a simple client-side adjustment) to make implementation and maintenance worth it,
- to having someone committed to implementing it and
- to having large enough demand for it.

(I don't want to argue in any direction whether we should have this feature or not, I just want to provide another perspective)

**Reply by pgScorpio** [reply]
> I understand that support for this scenario is beneficial in certain environments and it sounds like a simple feature or extension.

It's certainly beneficial for a lot of users, but I still don't call this a feature nor extension. It's simply correcting basic input/output selection.

> In fact, it might have been simple if it was built that way in the first place. 

Yes, I totally agree.

> However, it wasn't. My gut feeling is that this is not a trivial extension at all, 

I think it IS trivial. No code to add at all, just changing some checks and initialization. (See below)

> because it either requires a single Jamulus instance to behave like multiple instances on a protocol level (i.e. handling multiple streams) or because the protocol and/or server would require support for more than two channels in order to signal that back to clients. In any case, it would need UI changes.

No it doesn't. No change in Jamulus behavior at all when just changing basic input/output selections.. 
Multiple faders for multiple channels on a single client would be a nice extension though, but that's not what I am referring to. I just want always to be able to select the same input for both, left and right, channels or to swap left and right input/output channels. (Which is already possible, but only with audio interfaces with more than 2 channels)

> It certainly should be possible to build this.

Yes, I had some time yesterday evening, so I checked and modified the code for corrected input/output selection on Windows. 
Changed just 6 lines of code...  (in v3.8.1 release)

Added in sound.h:

```
42 #define MIN_IN_CHANNELS 1
43 #define MIN_OUT_CHANNELS 2

```
Chanced in sound.cpp:

```
168 if ( ( lNumInChan < MIN_IN_CHANNELS ) || ( lNumOutChan < MIN_OUT_CHANNELS ) )

586 vSelectedInputChannels[1]  = GetNumInputChannels() > 1 ? 1 : 0;

588 vSelectedOutputChannels[1] = GetNumOutputChannels() > 1 ? 1 : 0;

```
Chanced in clientsettingsdlg.cpp

```
823 if ( ( iNumInChannels < MIN_IN_CHANNELS  ) || ( iNumOutChannels < MIN_OUT_CHANNELS  ) )

```
And it works. Even with mono input devices.
I don't think there is any other impact on the GUI, but that has to be further checked/tested.

Same (alike) changes also have to be done for Mac/Linux/Android  sound.cpp/sound.h (and I''m actually amazed that this OS independent code is implemented in the OS/sound.cpp/.h in different ways... It's a pitty I currently don't find the time to look further into that, since it would make maintenance a lot easier.)

**Reply by dcorson-ticino-com** [reply]
I fail to see how your change has anything to do with the OP's comment.
Count me out.

**Reply by pgScorpio** [reply]
> I fail to see how your change has anything to do with the OP's comment.

I can understand your confusion, since it is only partly related to the original post. (Since on 2 channel interfaces you can't run two instances, each using one input. as suggested). unfortunately my remark started a whole new discussion that, as I mentioned multiple times, should not have been discussed in this topic.... Also my bad, since I probably should have opened new topic instead... Sorry for that.

EDIT: For those who want to discuss this futher: I opened an [issue](https://github.com/jamulussoftware/jamulus/issues/2344) for it.

**Reply by hoffie** [reply]
Heh, ok, that explains why what I said didn't match what you meant at all. Sorry, I've misread that as well as I was still primed with OP's post. :)
So, yeah, these are two different things:
1. *Support for multiple channels (with their own faders/names/mute status etc.) within a single client* (that's what OP wanted, as I understand, and what my reply was targeted at) vs.
2. *Support for channel selection*.

The latter might be a totally different story. I don't know any historical reasons for (not) having the latter, but if it's really that simple, doesn't break any other existing usage and would help people with such setups, than I guess it would be a welcome addition.

cc @softins @pljones

**Reply by henkdegroot** [reply]
> And it works. Even with mono input devices. I don't think there is any other impact on the GUI, but that has to be further checked/tested.
> 
> Same (alike) changes also have to be done for Mac/Linux/Android sound.cpp/sound.h

I don't think changes are need for other OS's, they do not have the same selection option as used on Windows. To my knowledge the input/output selection is only available in the dialog when using a Windows OS. Sounds like an issue needs to be created to get this improvement implemented.

**Reply by pgScorpio** [reply]
> I don't think changes are need for other OS's

AFAIK you do have input selection on macOS too, but I agree the input/output selection problem is mainly a windows (ASIO) problem.
Nevertheless the mono input restriction applies to all OS'es. 
And, in my opinion, the restrictions should not have been coded in the Sound module(s), since they are NOT sound module restrictions at all. 
EDIT: Correction on macOS the input/output selection is done via "Device", even stranger compared to the other OS'es. In my opinion this should be done via input/output selection here too.
